html/captive-portal/lib/captiveportal/PacketFence/Controller/TLSProfile.pm

```diff
@@ -0,0 +1,242 @@
+package captiveportal::PacketFence::Controller::TLSProfile;
+use Moose;
+use namespace::autoclean;
+use WWW::Curl::Easy;
+use Crypt::OpenSSL::PKCS12;
+use pf::log;
+use pf::config;
+use pf::util;
+use pf::node;
+#use pf::web qw(i18n ni18n i18n_format render_template);
+use pf::web::constants;
+use List::MoreUtils qw(uniq any);
+#use Readonly;
+#use POSIX;
+use pf::authentication;
+use HTML::Entities;
+#use URI::Escape::XS qw(uri_escape);
+use pf::web;
+use File::Basename;
+use Email::Valid;
+
+
+BEGIN { extends 'captiveportal::Base::Controller'; }
+
+__PACKAGE__->config( namespace => 'tlsprofile', );
+
+=head1 NAME
+
+captiveportal::PacketFence::Controller::TLSProfile - EAPTLS Controller
+
+=head1 DESCRIPTION
+
+Controller for EAPTLS connections.
+
+=head1 METHODS
+
+=cut
+
+=head2 index
+
+=cut
+
+sub index : Path : Args(0) {
+    my ($self, $c) = @_;
+    my $username = $c->session->{username};
+    my $logger  = get_logger;
+    my $profile = $c->profile;
+    my $request = $c->request;
+    my $mac = $c->portalSession->clientMac;
+    my $node_info = node_view($mac);
+    my $pid = $node_info->{'pid'};
+    my $provisioner = $c->profile->findProvisioner($mac);
+    #$provisioner->authorize($mac) if (defined($provisioner));
+    my $certificate_cn = $mac;
+    $certificate_cn =~ s/:/-/g;
+    $c->stash(
+        post_uri            => '/tlsprofile/cert_process',
+        certificate_cn      => $certificate_cn,
+        certificate_pwd     => $request->param_encoded("certificate_pwd"),
+        certificate_pwd_check     => $request->param_encoded("certificate_pwd_check"),
+        certificate_email   => lc( $request->param_encoded("certificate_email") || $request->param_encoded("email")),
+        template            => 'pki.html',
+        provisioner         => $provisioner,
+        username            => $username,
+        mac                 => $mac,
+        pid                 => $pid,
+    );
+}
+
+sub build_cert_p12 : Path : Args(0) {
+    my ($self, $c) = @_;
+    my $logger = $c->log;
+    my $session = $c->session;
+    my $cert_data = $c->stash->{'cert_content'};
+    my $cert = $c->stash->{'certificate_cn'} . ".p12";
+    my $certname = "$cert_dir/$cert";
+    $c->session( certificate_cn => "$cert" );
+    my $portalSession = $c->portalSession;
+    my $mac           = $portalSession->clientMac;
+    my $node_info     = node_view($mac);
+    my $pid           = $node_info->{'pid'};
+    my $fh;
+    open ($fh, '>', $certname);
+    if (-e $certname) {
```

Comment:
Instead of checking if it exists here, I would check if it exists and if it has content just after closing the file handle. This way you're more confident you have it right